### PR TITLE
HPCC-17711 Roxie ssl listener passphrase decrypt

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

testsocket and soapplus tests were run successfully

### CODE REVIEW

**Comment by mckellyln** [comment]
@afishbeck please review.

**Comment by mckellyln** [comment]
@afishbeck I had to change it so certfile, keyfile and passphrase are independent and persistent in case there were > 1 ssl listeners.
If you don't like the cast in free in dtor I can change these 3 to be not const.

**Comment by mckellyln** [comment]
@afishbeck updated again, removed cast and changed certfile, keyfile, passphrase to be not const.

**Comment by mckellyln** [comment]
@afishbeck updated again to use StringAttr.  Please review.

File: roxie/ccd/ccdprotocol.cpp [hunk 1]
```diff
@@ -283,13 +283,13 @@ class ProtocolSocketListener : public ProtocolListener
             Owned<ISecureSocket> ssock;
             if (client)
             {
-                if (protocol && streq(protocol, "ssl"))
+                if (protocol.get() && streq(protocol.get(), "ssl"))
                 {
```
Comment (mckellyln) [inline roxie/ccd/ccdprotocol.cpp]: Should I use .str() in the streq() ?

File: roxie/ccd/ccdprotocol.cpp [hunk 2]
```diff
@@ -283,13 +283,13 @@ class ProtocolSocketListener : public ProtocolListener
...
 #ifdef _USE_OPENSSL
                     try
                     {
                         if (!secureContext)
-                            secureContext.setown(createSecureSocketContextEx(certFile, keyFile, passPhrase, ServerSocket));
+                            secureContext.setown(createSecureSocketContextEx(certFile.get(), keyFile.get(), passPhrase.get(), ServerSocket));
```
Comment (mckellyln) [inline roxie/ccd/ccdprotocol.cpp]: Should I use .str() here ?

File: roxie/ccd/ccdmain.cpp [hunk 3]
```diff
@@ -1130,13 +1129,16 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
                 if (port)
                 {
                     const char *protocol = roxieFarm.queryProp("@protocol");
-                    const char *passPhrase = nullptr;
-                    const char *certFile = nullptr;
-                    const char *keyFile = nullptr;
+                    const char *certFilePtr = nullptr;
```
Comment (afishbeck) [inline roxie/ccd/ccdmain.cpp]: You don't need all of these xPtr variables.  Just use certFileName.str() when you pass it as a parameter to the createListener factory,

File: roxie/ccd/ccdmain.cpp [hunk 4]
```diff
@@ -1130,13 +1129,16 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
...
+                    StringBuffer certFileName;
+                    const char *keyFilePtr = nullptr;
```
Comment (afishbeck) [inline roxie/ccd/ccdmain.cpp]: Same here.

File: roxie/ccd/ccdmain.cpp [hunk 5]
```diff
@@ -1130,13 +1129,16 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
...
+                    StringBuffer keyFileName;
+                    const char *passPhrasePtr = nullptr;
```
Comment (afishbeck) [inline roxie/ccd/ccdmain.cpp]: same.

File: roxie/ccd/ccdprotocol.cpp (hunk 1 above)
Comment (afishbeck) [inline roxie/ccd/ccdprotocol.cpp]: yes, and if you do you don't need to first check protocol.get().
`if (streq(protocol.str(), "ssl"))`
is all you need.

File: roxie/ccd/ccdprotocol.cpp (hunk 2 above)
Comment (afishbeck) [inline roxie/ccd/ccdprotocol.cpp]: if createSecureSocketContextEx can handle nullptrs then this is fine.  If createSecureSocketContextEx can only handle "" then use .str()

File: roxie/ccd/ccdmain.cpp [hunk 6]
```diff
@@ -1167,7 +1174,7 @@ int STARTQUERY_API start_query(int argc, const char *argv[])
                     const char *soname =  roxieFarm.queryProp("@so");
                     const char *config  = roxieFarm.queryProp("@config");
                     Owned<IHpccProtocolPlugin> protocolPlugin = ensureProtocolPlugin(*protocolCtx, soname);
-                    roxieServer.setown(protocolPlugin->createListener(protocol ? protocol : "native", createRoxieProtocolMsgSink(ip, port, numThreads, suspended), port, listenQueue, config, certFileName.str(), keyFileName.str(), passPhrase));
+                    roxieServer.setown(protocolPlugin->createListener(protocol ? protocol : "native", createRoxieProtocolMsgSink(ip, port, numThreads, suspended), port, listenQueue, config, certFilePtr, keyFilePtr, passPhrasePtr));
```
Comment (afishbeck) [inline roxie/ccd/ccdmain.cpp]: Again, If you just do:
`createListener(protocol ? protocol : "native", createRoxieProtocolMsgSink(ip, port, numThreads, suspended), port, listenQueue, config, certFileName.str(), keyFileName.str(), passPhrase.str())`

Then you don't need all of those extra variables.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@mckellyln some comments.

**Comment by mckellyln** [comment]
@afishbeck updated.  Thanks for your comments.

**Comment by mckellyln** [comment]
Yes, createSecureSocketContextEx() is ok with nullptr args.

**Review by afishbeck — APPROVED** [review]
@mckellyln looks good

**Pull request merged**